habitat-lab/habitat/datasets/rearrange/viewpoints.py

```diff
@@ -0,0 +1,274 @@
+            return -1, pt, None, ViewpointType.down_unnavigable
+
+        pf = sim.pathfinder
+        pt = np.array(pf.snap_point(pt))
+        if pf.island_radius(pt) < ISLAND_RADIUS_LIMIT:
+            return -1, pt, None, ViewpointType.island_too_small
+        pt[1] += pf.nav_mesh_settings.agent_height
+
+        goal_direction = object_position - pt
+        goal_direction[1] = 0
+
+        q = direction_to_quaternion(goal_direction)
+
+        agent = sim.get_agent(0)
+        agent_state = agent.get_state()
+        agent_state.position = pt
+        agent_state.rotation = q
+        agent.set_state(agent_state)
+
+        cov = 0
+        for act_idx, act in enumerate(
+            [
+                "look_down",
+                "look_up",
+                "look_up",
+            ]
+        ):
+            agent.act(act)
+            obs = sim.get_sensor_observations(0)
+            cov += compute_pixel_coverage(obs["semantic"], semantic_id)
+
+            if debug_viz:
+                rgb_obs = np.ascontiguousarray(obs["color"][..., :3])
+                sem_obs = (obs["semantic"] == semantic_id).astype(
+                    np.uint8
+                ) * 255
+                contours, _ = cv2.findContours(
+                    sem_obs, cv2.RETR_LIST, cv2.CHAIN_APPROX_NONE
+                )
+                rgb_obs = cv2.drawContours(
+                    rgb_obs, contours, -1, (0, 255, 0), 4
+                )
+                img_dir = "data/images/objnav_dataset_gen"
+                os.makedirs(img_dir, exist_ok=True)
+                imageio.imsave(
+                    os.path.join(
+                        img_dir,
+                        f"{obj.handle}_{semantic_id}_{x}_{z}_{act_idx}.png",
+                    ),
+                    rgb_obs,
+                )
+
+        pt[1] -= pf.nav_mesh_settings.agent_height
+
+        keep_thresh = 0.001
+        if cov < keep_thresh:
+            return -1, pt, None, ViewpointType.low_visibility
+
+        return cov, pt, q, ViewpointType.good
+
+    candidate_poses_ious_orig = [_get_iou(*pos) for pos in candidate_poses]
+    poses_type_counter: Counter = Counter()
+    for p in candidate_poses_ious_orig:
+        poses_type_counter[p[-1]] += 1
+    candidate_poses_ious = [p for p in candidate_poses_ious_orig if p[0] > 0]
+
+    view_locations = [
+        ObjectViewLocation(
+            AgentState(pt.tolist(), quat_to_coeffs(q).tolist()), iou
+        )
+        for iou, pt, q, _ in candidate_poses_ious
+    ]
+    view_locations = sorted(view_locations, reverse=True, key=lambda v: v.iou)
+
+    if debug_viz and len(view_locations) == 0:
+        object_position_on_floor = np.array(object_position).copy()
+        if len(view_locations) > 0:
+            object_position_on_floor[1] = view_locations[
+                0
+            ].agent_state.position[1]
+        else:
+            while True:
+                # TODO: think of better way than ->
```

Comment:
Dont leave unassigned TODOs. Add your name here, so we can know who to ask when we merge this into main.

Reply:
I got this from Mukul, added his name.